python/lsst/ts/attcpip/csc.py

```diff
@@ -0,0 +1,292 @@
+        self._telemetry_task = utils.make_done_future()
+
+        # Keep track of the issued commands.
+        self.commands_issued: dict[int, CommandIssued] = {}
+
+        # Keep track of unrecognized telemetry topics.
+        self.unrecognized_telemetry_topics: set[str] = set()
+
+        # Iterator for command sequence_id.
+        self._command_sequence_id_generator = utils.index_generator()
+
+    @property
+    def connected(self) -> bool:
+        return self.cmd_evt_client.connected and self.telemetry_client.connected
+
+    async def handle_summary_state(self) -> None:
+        if self.summary_state in (salobj.State.DISABLED, salobj.State.ENABLED):
+            await self.start_clients()
+        else:
+            await self.stop_clients()
+
+    async def start_clients(self) -> None:
+        if self.connected:
+            self.log.warning("Already connected. Ignoring.")
+            return
+
+        # TODO DM-38912 Get this from the configuration.
+        host = ""
+        cmd_evt_port = 0
+        telemetry_port = 0
+
+        if self.simulation_mode == 1:
+            assert self.simulator is not None
+            await self.simulator.cmd_evt_server.start_task
+            await self.simulator.telemetry_server.start_task
+            host = self.simulator.cmd_evt_server.host
+            cmd_evt_port = self.simulator.cmd_evt_server.port
+            telemetry_port = self.simulator.telemetry_server.port
+
+        self.cmd_evt_client = tcpip.Client(
+            host=host, port=cmd_evt_port, log=self.log, name="CmdEvtClient"
+        )
+        self.telemetry_client = tcpip.Client(
+            host=host, port=telemetry_port, log=self.log, name="TelemetryClient"
+        )
+
+        await self.cmd_evt_client.start_task
+        await self.telemetry_client.start_task
+
+        self._event_task = asyncio.create_task(self.cmd_evt_loop())
+        self._telemetry_task = asyncio.create_task(self.telemetry_loop())
+
+    async def stop_clients(self) -> None:
+        if not self.connected:
+            self.log.warning("Not connected. Ignoring.")
+            return
+
+        self._event_task.cancel()
+        self._telemetry_task.cancel()
+
+        await self.cmd_evt_client.close()
+        await self.telemetry_client.close()
+
+        if self.simulation_mode == 1 and self.simulator:
+            await self.simulator.telemetry_server.close()
+            await self.simulator.cmd_evt_server.close()
+
+    async def close_tasks(self) -> None:
+        """Shut down pending tasks. Called by `close`.
+
+        Perform all cleanup other than disabling logging to SAL
+        and closing the dds domain.
+        """
+        await self.stop_clients()
+        await super().close_tasks()
+
+    async def cmd_evt_loop(self) -> None:
+        """Execute the command and event loop.
+
+        This loop waits for incoming command and event messages and processes
+        them when they arrive.
+        """
+        while True:
```

Comment:
If this runs forever, do you use the cancel function in async task to stop it?

Reply:
That is correct. That is exactly how it is stopped.